50 change management and git (#74)
phuse-org/git-in-statistical-programming: white-paper/Challenges.qmd

# Change Management and Git

## Introduction

> Statistical programmers in the pharmaceutical industry operate in a highly regulated environment where validated, reproducible analysis is paramount. For several years, the standard workflow has revolved around statistical computing environments managed through shared network drives, strict naming conventions, and manual version control practices (saving files with incremental version numbers or descriptive suffixes before archiving) or with regular server backups. These habits, while informal, have been deeply embedded in day-to-day practice and have served as the de facto audit trail in many organizations.

Reviewer: ```suggestion
Statistical programmers in the pharmaceutical industry operate in a highly regulated environment where validated, reproducible analysis is paramount. Historically, the standard workflow revolved around statistical computing environments managed through shared network drives, strict naming conventions, and manual version control practices (saving files with incremental version numbers or descriptive suffixes before archiving) or with regular server backups. These habits, while informal, have been deeply embedded in day-to-day practice and have served as the de facto audit trail in many organizations.
```

> Introducing Git into this context presents a unique set of challenges that go beyond mere tool adoption. It relies on notions such as sub-branches linked to a main branch, commits, remote repositories and merges, that can be difficult to adopt for teams used to working in other environments. The learning curve can be difficult and requires embracing a more collaborative and transparent workflow.

Reviewer 2: It introduces concepts such as branches, commits, remote repositories, pull requests, and merges that may be unfamiliar to teams accustomed to other environments.

> Introducing Git into this context presents a unique set of challenges that go beyond mere tool adoption. It relies on notions such as sub-branches linked to a main branch, commits, remote repositories and merges, that can be difficult to adopt for teams used to work in other environments. The learning curve can be difficult and requires to embrace a more collaborative and transparent workflow.

Reviewer: ```suggestion
Introducing Git into this context presents a unique set of challenges that go beyond mere tool adoption. It relies on notions such as sub-branches linked to a main branch, commits, remote repositories and merges, that can be difficult to adopt for teams used to working in other environments. The learning curve can be difficult and requires embracing a more collaborative and transparent workflow.
```

> Furthermore, the regulatory framework of the pharmaceutical industry is strict. Any change management system touching analysis code or submission-relevant outputs must be set in a GxP environment. Git must be used in a way that ensures compliance (traceability, audit, etc).

Reviewer 2: When Git is used as part of a process that creates or manages regulated records, its intended use and applicable GxP requirements should be defined through the organization’s quality system and a documented, risk-based assessment. Appropriate controls may include access management, traceability, review and approval, record retention, backup, and protection against unauthorized changes.

> Furthermore, the regulatory framework of the pharmaceutical industry is strict. Any change management system touching analysis code or submission-relevant outputs must be set in a GxP environment. Git must be used in a way that ensures theses compliances (traceability, audit, etc).

Reviewer: ```suggestion
Furthermore, the regulatory framework of the pharmaceutical industry is strict. Any change management system touching analysis code or submission-relevant outputs must be set in a GxP environment. Git must be used in a way that ensures compliance (traceability, audit, etc).
```

> Organizational culture also plays a significant role. Statistical programmers often work independently or in small teams with well-established personal workflows. Moving to a Git focused workflow requires not only technical training, but to being open to new mindsets and ways of working.

Reviewer: ```suggestion
Organizational culture also plays a significant role. Statistical programmers often work independently or in small teams with well-established personal workflows. Moving to a Git-focused workflow requires not only technical training, but to being open to new mindsets and ways of working.
```

## The Change Curve

> The Change Curve is a well known model for understanding how employees resond to change @HOLDSWORTH2024254. While there are many different implementations of it, generally speaking it is understood to be split into two main phases, first of resistance and then of acceptance. An important attitude in any company trying to make any sort of change, be it organisational or technological, is to have empathy for those individuals who are experiencing it, and give them the support they need.

Reviewer: I feel that it is likely better to use more than just two phases to describe the curve. Even better is if we can add an image depicting the curve.
```suggestion
The Change Curve is a well known model for understanding how employees respond to change @HOLDSWORTH2024254. While it has many different forms, it generally begins with initial shock, followed by resistance, low morale, gradual experimentation, and then acceptance and integration. An important consideration for any company trying to make any significant change, be it organisational or technological, is to have empathy for those individuals who are experiencing it and give them the support they need.
```

> The Change Curve is a well known model for understanding how employees respond to change @HOLDSWORTH2024254. While it has many different forms, it generally begins with initial shock, followed by resistance, low morale, gradual experimentation, and then acceptance and integration. An important consideration for any company trying to make any significant change, be it organisational or technological, is to have empathy for those individuals who are experiencing it and give them the support they need.

Reviewer: I imagine we don't want this "@"
```suggestion
The Change Curve is a well known model for understanding how employees respond to change. While it has many different forms, it generally begins with initial shock, followed by resistance, low morale, gradual experimentation, and then acceptance and integration. An important consideration for any company trying to make any significant change, be it organisational or technological, is to have empathy for those individuals who are experiencing it and give them the support they need.
```

Reviewer 3: ah this is how you make a citation, so I will keep this

> We must understand that any change will take time to embed into an organisation, and the main thing we can influence is how fast that change goes, and how well we can adopt the technology.

Reviewer: "and how well we can adopt the technology"--what does this mean?

Reviewer 3: I have reworded this, see what you think!

Reviewer: Looks good, thanks!

> In this chapter we will mostly ignore the how of using Git within your company (for more, please see the [recommendations](recommendations.qmd) chapter), and look instead about the things you can do to support your team as they learn to adopt Git in practice. 

Reviewer: ```suggestion
In this chapter, we will largely set aside the specifics of how Git is used within your organization (for more, please see the [recommendations](recommendations.qmd) chapter). Instead, we will focus on the ways you can support your team as they learn to adopt Git in their day-to-day work.
```

## Benefits of Git for statistical programmers

One important lesson from companies that have already experienced the transition to using Git is that, because Git can be a challenging technology to adopt at first, it can be easy to lose sight as to why this change is being made.

To combat this difficulty, we recommend highlighting early and often the benefits of adopting Git in practice. These are covered in more detail in the chapter on [benefits](benefits.qmd), but things you can easily show to someone who is learning Git:

> 1) The easy audit trail on a Github repository, and the ability to view code at different points in time trivially

Reviewer 2: GitHub needs to be written as "GitHub" instead of "Github."

2) Pull Requests, and the ability to review code line by line and provide easy feedback

> 3) AI Integrations which can improve code review in practice

Reviewer 2: Approved AI-assisted tools may support code review, subject to organizational policies, data-protection requirements, and appropriate human review.

Reviewer 3: I will include these more in the benefits section

4) Findability and searchability of code

> 

Reviewer 2: The following benefits can be demonstrated early in the adoption process:
1) A traceable version history that allows programmers to compare code at different points in time.
2) Pull requests that support line-by-line peer review and documented discussion.
3) Parallel development through branches without overwriting another programmer’s work.
4) Improved searchability and reuse of code across repositories.
5) Integration with automated checks such as unit tests, package checks, and style checks.

All of these are very easy to show quite quickly, and should not be forgotten when trying to bring someone into a new technology

## Introducing tools

> Depending on your company's size and how they want to use Git, there a number of tools to interact with Git and Github. These are covered more in the [appendix](tools.qmd), but your decision is likely to be motivated based on how close you want users to interact with Git.

Reviewer 2: there a number → there are a number

> You may decide that you wish to abstract users access to Git or you might decide that you wish to get users comfortable with direct Git terminal access. Different tools will enable different approaches.

Reviewer: Is this in line with the intention of this sentence? 
```suggestion
You may choose to hide Git's complexity from users by giving them easy-to-use tools, or you may choose to teach users to work directly with Git commands in the terminal. The tools you select will influence which approach is possible.
```

Note that of course you are not limited to using any one tool, and can make multiple available to support different workflows and different levels of ability in your user base. The important thing when it comes to managing change is to make it clear to individuals what tools are for, and when they are using them. For consistency, it may be wise to limit teams that work together to at most two common tools. 

## Preparing Documentation

There are a number of excellent resources for learning Git online. This means that when creating internal documentation on Git it is important to focus on how your organisation will be using Git in practice.

> Your documentation should focus first on the ways someone will work in their day to day, covering the typical flow and where possible integrating it into the tools they are likely to use. Documentation should be easy to access and use, and easy to search. We recommend where possible to not include this level of documentation in GxP training, as the latter is likely to be in systems that can be harder to update quickly based on user feedback.

Reviewer: ```suggestion
Your documentation should focus first on the ways someone will work in their day-to-day, covering the typical flow and where possible integrating it into the tools they are likely to use. Documentation should be easy to access and use, and easy to search. We recommend where possible to not include this level of documentation in GxP training, as the latter is likely to be in systems that can be harder to update quickly based on user feedback.
```

> GxP training is of course important to meet our compliance requirements, but we can normally build training on top of it which enables more dyanmic updates as you get feedback from users on what works and what does not. 

Reviewer: ```suggestion
GxP training is of course important to meet our compliance requirements, but we can normally build training on top of it which enables more dynamic updates as you get feedback from users on what works and what does not. 
```

> ### Failure cases

Reviewer 2: Troubleshooting common scenarios

> An important resource to build over time is how to resolve issues that might occur when using Git. The exact nature of the issues are likely to run into will be specific to your organisations workflow but some likely common issues are

Reviewer: ```suggestion
An important resource to build over time is how to resolve issues that might occur when using Git. The exact nature of the issues users are likely to run into will be specific to your organisation's workflow but some likely common issues are
```

- Merge conflicts and how to resolve them smoothly
- Undoing things (git reset and revert)

> - Not being able to switch brances (git stash)

Reviewer 2: brances → branches

> 

Reviewer 2: Resolving merge conflicts safely
Reverting a change in shared history using git revert
Rewriting unshared local history using git reset, with appropriate caution
Handling uncommitted changes before switching branches using git stash or a work-in-progress commit
Resolving authentication and repository-permission issues

Having simple, easy to understand guides to help in these cases can be very useful. These are things that can be taught, but don't come up frequently enough for the knowledge to necessarily be embedded, so having clear instructions on these can greatly improve Git adoption.

## Practical examples as training

When teaching users to use Git, we should always focus on practical examples which relate as closely to how the user will be applying Git in practice. While for fundamentals it can be helpful to be artificial, as soon as your users have grasped the basic, we strongly encourage examples that involve practicing the workflow they do in person.

Where possible trainings should be live, and in person if it is possible to arrange, to allow maximum engagement.

> One vital aspect is that training should be targetted only at individuals who are going to be adopting the Git workflow very soon, or who are already adopting it. This means that what they have learnt can be applied straight away, removing the danger of knowledge atrophying during use. Related to this, if you can train cohorts of users who will be working together, this can be very useful as they will be able to learn and practice together. 

Reviewer: ```suggestion
One vital aspect is that training should be targetted only at individuals who are going to be adopting the Git workflow very soon, or who are already adopting it. This means that what they have learnt can be applied straight away, removing the danger of knowledge deteriorating prior to use. Related to this, if you can train cohorts of users who will be working together, this can be very useful as they will be able to learn and practice together. 
```

> One vital aspect is that training should be targetted only at individuals who are going to be adopting the Git workflow very soon, or who are already adopting it. This means that what they have learnt can be applied straight away, removing the danger of knowledge atrophying during use. Related to this, if you can train cohorts of users who will be working together, this can be very useful as they will be able to learn and practice together. 

Reviewer 2: targetted → targeted

## Feedback and checking in

One important way to manage change in an organisation is to demonstrate that you are open to feedback, and responding to it. This means you must be actively seeking out feedback, not just passively waiting for it. Consider small focus groups, arranging meetings with key thought leaders, alongside regular surveys of sentiment. 

When taking feedback, it is important to understand the context in which it is delivered. Try to not be too reactive: making changes immediately may not be the best course of action, but try to fully understand what the individual is saying and how you can best address it. In some cases, it may truly be that the Git workflow your organisation has chosen to adopt is not fit for purpose, but in others it may be that you need better training and support to help users address scenarios they do not currently understand.

When you do make changes based on feedback, make sure you communicate them clearly, and make clear the reason you chose to adopt them.

## Conclusion

> Any kind of change can have large impacts on an organisation, and Git as a technology can definitely cause some friction. Just like with all changes, provided it is handed well and responsively, then within a few years the use of Git will simply be the new normal. 

Reviewer 2: When the transition is handled well and supported in a responsive way, Git can become a routine and sustainable part of the statistical programming workflow.

> Any kind of change can have large impacts on an organisation, and Git as a technology can definitely cause some friction. Just like with all changes, provided it is handed well and responsively, then within a few years the use of Git will simply be the new normal. 

Reviewer 2: handed → handled
